### PR TITLE
Fix failing smoke tests when a new release candidate is published

### LINKED ISSUE #31833: Investigate why smoke tests for WooCommerce release candidates fail

### PR DESCRIPTION
**Pull request by MrJnrman**
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?
### Changes proposed in this Pull Request:

The first step in publishing is to use `Woorelease` which creates a release on GitHub. Next, the `build-release.yml` workflow gets triggered, which packages and uploads a `woocommerce.zip` asset for that release. 

The reason why the `smoke-test-release.yml` workflow [fails](https://github.com/woocommerce/woocommerce/runs/5054386785?check_suite_focus=true) is because it gets triggered on release publish events, and in this case, it wouldn't have an asset to download until after the `build-release.yml` workflow is complete.

The fix proposed is to have the `build-release.yml` workflow trigger the `smoke-test-release.yml` after it has uploaded the new asset. That way `smoke-test-release.yml` won't run into the issue of having a release with no asset attached to the issue.

Things to note:
- This workflow adds `E2E_WORKFLOW_GH_TOKEN` to our secrets. A reminder has been set up to prompt the team to update the token a day before it expires. 
- The `workflow_dispatch` changes won't be picked up by GitHub until this branch is merged(workflows with that event trigger must be on the default branch to be recognized by GitHub)
- Since this depends on a RC to be published, it may make sense to have a test PR after merging to ensure this implementation works as expected.

Closes #31833.

### How to test the changes in this Pull Request:
I've created a public repository at https://github.com/MrJnrman/woo-workflow-test to test the core of these changes. Its basically a copy-paste of the logic to trigger the `build-release.yml` on the publishing of a release, which subsequently triggers a workflow that runs the `Download WooCommerce release zip` workflow step.

1. Clone or fork the repository
2. [Generate](https://github.com/settings/tokens) an access token with permissions for `workflows`
3. Create a project secret called `CUSTOM_GH_TOKEN` with its value set to the generated access token
4. Create the following releases and observe the workflows from the `Actions`:
   - A pre-release with an asset
   - A pre-release without an asset
   - A release with an asset
   - A release without an asset

**Note: Do not upload a zip with the name `woocommerce.zip` this will break the workflow**

### CODE REVIEW

File: .github/workflows/scripts/fetch-asset-id.js [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+ */
+const https = require('https');
+
+const options = {
+	hostname: 'api.github.com',
+	port: 443,
+	path: `/repos/${process.env.REPO}/releases/${process.env.RELEASE_ID}/assets`,
+	method: 'GET',
+	headers: {
+		'Content-Type': 'applicatipon/json',
```
Comment (Luc45) [inline .github/workflows/scripts/fetch-asset-id.js]: applicatipon

File: .github/workflows/smoke-test-release.yml [hunk 2]
```diff
@@ -72,6 +75,17 @@ jobs:
...
+      - name: Fetch Asset ID
+        id: fetch_asset_id
+        uses: actions/github-script@v5
+        env:
+          RELEASE_ID: ${{ github.event.inputs.release_id }}
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+          REPO: ${{ github.repository }}
+        with:
+          script: |
+            const script = require( './package/woocommerce/.github/workflows/scripts/fetch-asset-id.js' )
```
Comment (Luc45) [inline .github/workflows/smoke-test-release.yml:87]: I didn't know you could do that in GitHub workflows :exploding_head: :thinking:

File: .github/workflows/scripts/fetch-asset-id.js (hunk 1 above)
Comment (MrJnrman) [reply .github/workflows/scripts/fetch-asset-id.js]: Thanks for catching this! I feel like it shouldn't have worked with that typo :')

File: .github/workflows/smoke-test-release.yml (hunk 2 above)
Comment (MrJnrman) [reply .github/workflows/smoke-test-release.yml:87]: I really like this feature. It gives us a ton of flexibility

**Comment by Luc45** [comment]
Howdy @MrJnrman , thanks for working on this!

I followed a slightly different testing instruction, by forking `woocommerce/woocommerce` repository itself and replacing all GitHub workflows with the ones in https://github.com/MrJnrman/woo-workflow-test

See: https://github.com/Luc45/woocommerce/tree/trunk/.github/workflows

When I published a release, it did download woocommerce.zip as expected, but it did not run the Smoke Test: https://github.com/Luc45/woocommerce/actions/runs/1829939856

Is it expected that it just download the zip but not run the smoke test, or am I doing something wrong?

**Comment by Luc45** [comment]
It seems the file that is being downloaded is a development version of the repository, I believe we should be downloading the distributable package of `woocommerce.zip` for the smoke test? The same that is shipped to the customers?

https://github.com/Luc45/woocommerce/runs/5157826624?check_suite_focus=true

**Comment by MrJnrman** [comment]
Thanks for looking into this @Luc45 

> Is it expected that it just download the zip but not run the smoke test, or am I doing something wrong?

This PR address the workflow not being able to download a zip when `Woorelease` automatically triggers a release event with no asset attached. So the changes in my repo only covered those specific changes instead of actually running the entire test suite.

> It seems the file that is being downloaded is a development version of the repository, I believe we should be downloading the distributable package of woocommerce.zip for the smoke test? The same that is shipped to the customers?

The workflow being modified has no issues with stable releases which is why the focus here is on release candidates. That said, these changes work for stable or pre-releases, with or without assets. When testing, you can try all four combinations to confirm.

**Comment by Luc45** [comment]
Wrapping up what we discussed in DM in Slack just so it doesn't get lost in time, we should check if the downloaded zip should be of the WooCommerce plugin release _(the same that you can upload to a WordPress website to install it)_, or the WooCommerce GitHub repo _(the one we clone with git clone)_.

This PR seems to download the repo, and I think there's a chance that we need the release version instead.

**Comment by MrJnrman** [comment]
For clarification, this PR fixes the downloading of the asset attached the the [Pre-Release/RC](https://github.com/woocommerce/woocommerce/releases) which is created by the [build-release workflow](https://github.com/woocommerce/woocommerce/blob/trunk/.github/workflows/build-release.yml). That said, I intend to investigate if it is intentional to have our Pre-Releases be built differently than our stable releases. The rationale behind that question is that we want to ensure that we are testing a build as close to production-like as possible.

**Review by rrennick — APPROVED** [review]
@MrJnrman Thanks for getting this one. As mentioned in the PR we'll have to follow up and verify on the next pre-release.

**Pull request merged**